builtin/logical/mongodb/path_roles.go

```diff
@@ -0,0 +1,182 @@
+			},
+			"roles": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "MongoDB roles to assign to the users generated for this role.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("role/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	rolesJsonBytes, err := json.Marshal(role.MongoDBRoles)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"db": role.DB,
+			"roles": string(rolesJsonBytes),
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	roleDB := data.Get("db").(string)
+
+	// Example roles JSON: [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	var roles []interface{}
+	rolesJson := []byte(data.Get("roles").(string))
+	if len(rolesJson) > 0 {
+		var rolesArray []interface{}
+		err := json.Unmarshal(rolesJson, &rolesArray)
+		if err != nil {
+			return nil, err
+		}
+		for _, rawItem := range rolesArray {
+			switch item := rawItem.(type) {
+			case string:
+				roles = append(roles, item)
+			case map[string]interface{}:
+				if db, ok := item["db"].(string); ok {
+					if r, ok := item["role"].(string); ok {
+						roles = append(roles, mongodbRole{Role: r, DB: db})
+					}
+				}
+			}
+		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON("role/"+name, &roleEntry{
+		DB:    roleDB,
+		MongoDBRoles: roles,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+type roleEntry struct {
+	DB           string        `json:"db"`
+	MongoDBRoles []interface{} `json:"roles"`
```

Comment:
> I don't think we can send the following to Mongo:
> [ { "role": "readWrite" } ]

We don't need to though. We can create the correct objects before serialization.

> I had opted against this option because it amounts to reimplementing the behavior of Mongo's createUser command, and if Mongo changes what it does with bare role names in the roles array down the road, we'd have to update Vault to match.
> What do you think?

I think it's unlikely that they will change this. I think it's more likely that they will add the ability to specify more options in the dict version.

> An additional consideration: do we want the roles JSON produced by GET /mongodb/roles/<name> to match what the user provided when they created the role?

Ideally, but not at the expense of difficult code.

Out of these options, I think my preferred choise is number 2 -- the one you opted against :-)  Doing this will produce a correct object that can be fed directly into Mongo, and although it might be different than what the user gave, it will also be technically correct and usable (e.g. if they then want to paste into the Mongo CLI).

However, you mention that we will know the user's auth DB. Is this the same DB as is given in the `db` parameter of this path? If so, if this endpoint ever respects the difference between `Create` and `Update` then codifying the DB will mean updating the DB parameter alone won't have the desired effect. One easy way to avoid this is to simply never have a difference between the two :-)

I do think the whole concept of roles in different DBs is pretty odd. `createUser` is a method scoped to a DB, and with the user recorded in that DB, so then assigning roles valid on other DBs feels...weird. Weird enough that I'd be totally happy simply dropping support for that altogether here, and only allowing roles corresponding to the given `db` parameter. If clients also want creds for a different DB, they could read them against a different role.